torch/ao/quantization/fx/_model_report/model_report.py

```diff
@@ -0,0 +1,105 @@
+from typing import Dict, Set, Tuple
+
+import torch
+from torch.ao.quantization.fx._model_report.detector import DetectorBase
+from torch.ao.quantization.fx.graph_module import GraphModule
+
+
+DEFAULT_MODEL_REPORT_OBSERVER_PRE_EXTENSION = ".model_report_pre_observer"
+DEFAULT_MODEL_REPORT_OBSERVER_POST_EXTENSION = ".model_report_post_observer"
+
+
+class ModelReport:
+    r"""
+    Generates report and collects statistics
+        Used to provide users suggestions on possible model configuration improvements
+
+    Currently supports generating reports on:
+    - Suggestions for dynamic vs static quantization for linear layers (Graph Modules)
+
+    * :attr:`desired_report_detectors` The set of Detectors representing desired reports from the ModelReport class
+        Make sure that these are all unique types of detectors [do not have more than 1 of the same class]
+
+    Proper Use:
+    1.) Initialize ModelReport object with reports of interest by passing in initialized detector objects
+    2.) Prepare your model with prepare_fx
+    3.) Call model_report.prepare_detailed_calibration on your model to add relavent observers
+    4.) Callibrate your model with data
+    5.) Call model_report.generate_report on your model to generate report and optionally remove added observers
+
+    """
+
+    def __init__(self, desired_report_detectors: Set[DetectorBase]):
+
+        if len(desired_report_detectors) == 0:
+            raise ValueError("Should include at least 1 desired report")
+
+        # keep the reports private so they can't be modified
+        self.__desired_report_detectors = desired_report_detectors
+        self.__desired_reports = set([detector.get_detector_name() for detector in desired_report_detectors])
+
+        # keep a mapping of desired reports to observers of interest
+        # this is to get the readings, and to remove them, can create a large set
+        # this set can then be used to traverse the graph and remove added observers
+        self.__report_to_observers_of_interest: Dict[str, Set[str]] = {}
+
+        # initialize each report to have empty set of observers of interest
+        for desired_report in self.__desired_reports:
+            self.__report_to_observers_of_interest[desired_report] = set([])
+
+    def get_desired_reports(self) -> Set[str]:
```

Comment:
This method name makes it sound like it's returning the actual report (with the detailed dict), but it's actually just returning a list of detector names. Do we need this method?

Reply:
I changed the name so it better reflects the fact that it's returning the names. I think it's definately useful so that a user can see which detectors they had added since they cannot access the detectors (so they don't accidently modify them). I think it's worth keeping for now since it doesn't hurt, and we can remove it later as I flesh out the system more if I realize that it's really not necessary.